spec/ddtrace/logger_spec.rb

```diff
@@ -0,0 +1,244 @@
+require 'spec_helper'
+
+require 'logger'
+require 'stringio'
+require 'time'
+require 'ddtrace'
+
+RSpec.describe Datadog::RateLimitedLogger do
+  # Define functions to log to ensure the caller stack is consistent between tests
+  def log_warn
+    Datadog::Tracer.log.warn('warn message')
+  end
+
+  def log_error
+    Datadog::Tracer.log.error('error message')
+  end
+
+  def log_debug
+    Datadog::Tracer.log.debug('debug message')
+  end
+
+  # Ensure we restore original values after every test
+  around do |example|
+    Datadog::Tracer.log = Datadog::Tracer.log.tap do
```

Comment:
Sorry I'm only thinking about this now, but wouldn't it make more sense to just test an instance of `Datadog::RateLimitedLogger`, instead of setting and resetting the global logger?

These are unit tests and should probably focus only on testing the behavior of `Datadog::RateLimitedLogger`, not how it's plugged into the the global configuration.

I might suggest some changes like replacing `around` with:

```
# This is basically what you had before, just as a subject instead
subject(:logger) do
  described_class.new(Datadog::Logger.new(buf)).tap do |logger|
    logger.level = log_level
  end
end

# Set this in your examples, where appropriate.
let(:rate_setting) { 0 }

# Stub global configuration instead of changing global configuration
before(:each) { allow(Datadog.configuration.logging).to receive(:rate).and_return(rate_setting)
```

Then you can basically replace all references to `Datadog::Tracer.log` with `logger`.

Reply:
yeah, this makes sense. I was refactoring some of the tests recently and realized I was just fighting so hard to make it work with the global tracer logger (assumed we wanted that because the old tests did), but I'll change these ones to only test a fresh instance.